### PR TITLE
111 allow user to select asset for fee

### LINKED ISSUE #111: [19] #1117 Allow a user to specify an arbitrary asset for transaction fee
https://github.com/bitshares/bitshares-ui/issues/1117

Description of the task is here https://github.com/bitshares/bitshares-ui/issues/1117#issuecomment-492209474

Please make sure that 
- [ ] Modal is only rendered when opened
- [ ] Unnecessary re-renders are avoided

### PR DESCRIPTION
**Pull request by Metizik**
<h2>General</h2>
Closes #111 

Refactored SendModal and extracted FeeAssetSelector component with all logic related to fee calculations and managing assets to use.
Implemented SetDefaultFeeAssetModal, which allows to select Asset to pay fee and update settings as described in issue comments.
Added settings entry for default fee asset selection and integrated it with new modal.
Integrated default fee asset settings into Fee Schedule view of the wallet.

<h2>General</h2>
Please make sure the following is done:

- Pull request is onto develop
- If you haven't already, have a look at
  - https://github.com/bitshares/bitshares-ui/blob/develop/CODE_OF_CONDUCT.md
  - https://github.com/bitshares/bitshares-ui/blob/develop/CONTRIBUTING.md

<h2>Code Preparation</h2>

_Please review all your changes one last time before committing_

- [x] Check for unused code
- [x] No unrelated changes are included
- [x] None of the changed files are reformatting only
- [x] Code is self explanatory or documented
- [ ] All written text is properly translated (english language)

<h2>Testing</h2>

_The branch has been tested on the following browsers (desktop and mobile view)_

- [x] Chrome 
- [ ] Opera
- [x] Firefox
- [ ] Safari

<h2>User interface changes</h2>

_Delete this section if there weren't any UI changes. Please make sure you tested your changes in all themes_

- [x] Dark
- [x] Light
- [x] Midnight

_Please provide screenshots/licecap of your changes below_

Screenshot from 2019-06-07 14-55-32
Screenshot from 2019-06-07 14-55-43
Screenshot from 2019-06-07 14-55-53
Screenshot from 2019-06-07 14-56-09

### CODE REVIEW

**Comment by sschiessl-bcp** [comment]
- please reference the issue of our repository on a PR and not the one in bitshares
- please use a styleguide table in "Select asset to pay fee" modal for consistency, styling can be default
- please display the chosen default fee asset in the settings component, with the button next to it

Will do integration test and code review on monday

File: app/components/Utility/FeeAssetSelector.jsx [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+            fee_asset_id:
+                ChainStore.assets_by_symbol.get(
+                    props.settings.get("fee_asset")
+                ) || "1.3.0",
+            fees: {},
+            feeStatus: {},
+            isModalVisible: false,
+            error: null
+        };
+        this._updateFee = debounce(this._updateFee.bind(this), 250);
```
Comment (sschiessl-bcp) [inline app/components/Utility/FeeAssetSelector.jsx:35]: Should we rather debounce the whole component instead of those two calls?

**Review by sschiessl-bcp — CHANGES_REQUESTED** [review]
Please checl the todo's in FeeAssetSelector.jsx

**Comment by sschiessl-bcp** [comment]
See here as well, starting https://github.com/blockchainprojects/bitshares-ui/issues/111#issuecomment-502655596

File: app/components/Utility/FeeAssetSelector.jsx (hunk 1 above)
Comment (mettizik) [reply app/components/Utility/FeeAssetSelector.jsx:35]: We should debounce actualy only updateFee, so I believe it will be ok to leave this one

**Comment by mettizik** [comment]
@sschiessl-bcp , all TODOs and referenced issues are closed now, if there are no other change requests I believe this request can be synced with actual develop branch and merged

**Comment by sschiessl-bcp** [comment]
You can try my account on mainnet (no login necessary)
http://localhost:8080/account/sschiessl

When the modal is opened, the default asset for fee and the fee amount is loaded and then the remaining assets of the user account queried and displayed in the fee asset select dropdown.

I want to change the logic: Show the default asset and asset fee as is now, but only load other choices lazily when the user clicks on the dropdown.
Following reasoning:
- When I first open the modal and don't do anything, it does heavy amount of re-renders (probably due to loading all my assets and check them for fee selection). 
- The dropdown does not reliable get editable after loading is complete. I must manually enter something in the memo, which triggers fee re-calculation, and then the dropdown becomes unlocked
- I expect 95% of users being fine with the default, thus that is unnecessary effort

Ideally, the modal loads the default and the checkbox is clickable. When the user click, a loading indicator is shown and when done the dropdown expands.

**Comment by sschiessl-bcp** [comment]
image

The select modal also doesnt seem to load properly when opened from within the new component. Select e.g. BADCOIN from my account

**Comment by Metizik** [comment]
@sschiessl-bcp , I fixed all the places to use default fee asset from settings appropriately.
Left to do is lazy loading of available assets, which will be done today.

**Comment by Metizik** [comment]
@sschiessl-bcp , all stated tasks are done. 
Regarding lazy load - I did not find a way to implement lazy loading like you proposed with `AssetSelect` component, but I removed all redundant blockchain fetching made it lightweight enough IMO, to address issue you mentioned. 
We have a few modals left that could be refactored to use new `FeeAssetSelector`. Should those be refactored too?

**Comment by sschiessl-bcp** [comment]
> @sschiessl-bcp , all stated tasks are done.
> Regarding lazy load - I did not find a way to implement lazy loading like you proposed with `AssetSelect` component, but I removed all redundant blockchain fetching made it lightweight enough IMO, to address issue you mentioned.
> We have a few modals left that could be refactored to use new `FeeAssetSelector`. Should those be refactored too?

Which modals are you referencing that could be refactored?

**Comment by sschiessl-bcp** [comment]
Please state which modals are missing. I will forward this PR now to not get out of touch with codebase too much. We can use new issue to streamline usage.

**Comment by Metizik** [comment]
@sschiessl-bcp , modals I am talking about:

1. app/components/Dashboard/SimpleDepositWithdraw.jsx
2. app/components/Showcases/Barter.jsx
3. app/components/Modal/HtlcModal.jsx
4. app/components/Modal/WithdrawModalNew.jsx
5. app/components/Modal/DirectDebitModal.jsx
6. app/components/Modal/DirectDebitClaimModal.jsx

**Pull request closed without merging**